### ISSUE TITLE
Resource versioning - creating new versions of resources

### ISSUE THREAD

**Issue by hyi** [hydroshare/hydroshare#946]
See design specification https://docs.google.com/document/d/1tNFStpEmbB_AC2ICppVSw-024-iTrnqGK9Np7Elw2OM/edit for details on resource versioning including workflow for creating new versions of unpublished and published resources.

**Comment by hyi** [comment]
@horsburgh To represent relationships between a new version and the original resource, the design spec says two pairs of terms "isVersionOf" and "isReplacedBy" and "obsoletes" and "isObsoletedBy" can be used. I am wondering whether one pair "isVersionOf" and "isRelacedBy" is sufficient since "isReplacedBy" can also mean "obsolete" relation. Currently, "isVersionOf" is the only relation term already implemented, so I will have to add other terms as needed for resource versioning and feel like adding the reverse relation term "isReplacedBy" to the existing "isVersionOf" appears to be sufficient to represent the relationship between a new version and the original resource so that I don't have to add one additional pair of relation terms. Let me know what you think.

**Comment by hyi** [comment]
@horsburgh I am using "isVersionOf" and "isReplacedBy" pair to show releationships between a new version and the obsoleted resource, and display a link to the obsoleted version on the new version's landing page indicating this resource obsoletes the resource with a link, and display a link to the newer version on the obsoleted version's landing page indicating this resource is obsoleted by the resource with a link. On "Related Resources" tab, "isVersionOf" and "isReplaceBy" are displayed with a link to corresponding resource. This appears to suffice to serve its purpose. Let me know if you want me to change anything regarding current implementation.

**Comment by horsburgh** [comment]
@hyi - Given that we have decided that versioning chains cannot split and that a new version always replaces the older version, the terms "isVersionOf" and "isReplacedBy" are adequate and we do not need the "obsoletes" and "isObsoletedBy" terms. 

Before, we were considering allowing versioning chains to split and so it would be possible to have two resources that were both newer versions of an existing resource and only one (or neither of them) might "obsolete" the other resource.  Having linear versioning chains simplifies this and so I think we can use just the two terms.

For display purposes on the landing page, I would change the language and not use the terms "isReplacedBy" and "isVersionOf".  I would use the following:

* A newer version of this resource is available that replaces this version (use this for isReplacedBy)
* An older version of this resource is available (use this text for isVersionOf)

You could make the "newer version" and "older version" text hyperlinks to the other resources or just provide the URL to the other resources.

**Comment by horsburgh** [comment]
From @hyi:

While working on resource versioning, one question comes up: should we allow versioning for web app or tool resource type and reference time series resources? I can kind of see the value for allowing to create new version of a reference time series resource, but want to confirm with you to be sure. I cannot see the utility for allowing to create new version of a web app or tool resource, so if you think we should allow it, could you briefly explain a use case to create a new version for a tool resource?

**Comment by horsburgh** [comment]
From @dtarb:

I think that versioning should be the same for all resources reference and physical just for simplicity and uniformity.  Apps will certainly be improved, so we will get new versions.  I am writing this email on Outlook version X.  The question is whether HydroShare needs to keep track of them and while the need for this is hard to anticipate I think that the same capability for all resources may find uses for apps.

**Comment by horsburgh** [comment]
@danames - can you review the comments above from @hyi and @dtarb about whether HydroShare needs capability for Apps and referenced resources to be versioned?  If you could weigh in, I think Hong can move forward.  Thanks!

**Comment by hyi** [comment]
@horsburgh @dtarb According to the design spec, we want to remove obsoleted resources from search. I am wondering whether it is a good idea to remove obsoleted published resources from search. Since we display an informational message indicating a newer version of this resource is available with a link to its new version, perhaps it still makes sense to keep them in the search if the obsoleted resources have a sharing status published, public, or discoverable? I feel like it makes things simpler if we don't make an exception to sharing status of obsoleted resources and make them still searchable. Let me know what you think.

**Comment by horsburgh** [comment]
@hyi - a resource that has been replaced by a newer version is obsolete.  It still needs to be available in HydroShare if someone has cited it. But, when someone is searching, I think we should only expose the most recent resource in an obsolescence chain. They can trace back in time to previous versions if they want by clicking the links on the resource landing page to the previous version, but removing previous versions from search avoids the confusion of people finding multiple versions of the same resource and not knowing which one to use.

I don't have a feel for how hard it would be to remove obsoleted resources from search though.  Maybe you could implement piecewise - get versioning of resources working and then add the bit to remove from search later?

**Comment by dtarb** [comment]
The thoughts I have on this are:

1.  I agree with @horsburgh that it would be better if searches return only the most recent version. By definition versions refer to different versions of the same thing, so searches one return that is in the search result listing that goes to the latest version landing page would be simplest for users.  However I am supportive of the simplicity of keeping versioning, search and sharing status all separate.  I think this can be done using the landing page.  If one goes to a landing page for a resource (from a search return or anywhere) it should provide links (perhaps via a listing) to earlier (and later) versions and an indication of whether the resource is the most recent version.  I think that this is sufficient for our first implementation and is all that should be the responsibility of the current work on versions.  This does imply a requirement to have other versions listed on a landing page.  

2.  I think that changes to search should be a separate issue, addressed by the search developers that should not delay versioning implementation.  (A search return that identifies multiple versions of the same resource could be filtered to return the latest only by the search function, but this is the work of the search function developers, not versioning).  I also think that versioning should not touch the public, published, sharing or access permissions functionality.   

3.  I realize that there may be complications associated with resources that are published and get new versions.  A DOI may point to a landing page of not the latest version.  This is OK.  Early version landing pages should include links to more recent versions.

4.  I realize that there may be complications associated with access control as different versions may have different access permissions.  I think we need to accept this and let landing page and access control deal with it.  We are establishing the policy that resource title and owners are always public so any resource landing page should present this information and a user can ask the owner for access if they do not have permission.  A feature to add to such a landing page for a resource that has earlier versions that may be public or shared could be links to those earlier versions.

I do worry that this discussion can go on and on, and that this is likely not something we are going to get perfect at the beginning and will need to be refined over time, so I would prefer to actually limit puzzling over this too much and just get an implementation that does number 1 above and address the other points later and in response to user feedback.

**Comment by hyi** [comment]
@horsburgh @dtarb Agreed it'd be better to get versioning functionality out there first, and refine it over time according to user feedback and use cases. Since @alvacouch and @xuezhaokun are revamping the current search and on-demand solr search update is not implemented yet, it is definitely better to remove all versions from searches as a separate issue to be addressed later after search is more stabilized with on-demand search update implemented. In fact, I have addressed the last remaining issue on Friday for versioning that locks the original resource when a new version of it is being created to guarantee a single obsolesence chain if we want to address search removal for obsoleted resources later. However, one pending issue as mentioned earlier needs weigh-in from @danames regarding whether we allow resource versioning for reference time series resources and web app resources. Since @dtarb prefers to allow this just for consistency across all resource types, I am leaning toward implementing it, but it'd be great if I can hear from @danames to confirm there is utility to do versioning for reference time series resources and web app resources. I am planning to implement versioning for these two special resource types on Monday unless I hear opposing views from @danames. Then I will deploy it on a VM for testing and feedback and open a PR for code review as well.

**Comment by dtarb** [comment]
Part of the motivation behind my suggestion for versioning functionality to be the same for all resources is that its implementation can be on the base resource class without any special handing of any resource types.  Your comment @hyi about "implement versioning for these two special resource types" makes me worried that there is separate code being written for separate resource types, which I would prefer not to be the case.

**Comment by hyi** [comment]
@dtarb Unfortunately, this is indeed the case due to the complexity of our current metadata structure. BaseResource metadata model handles core science metadata copying from source resource to target (new versioned) resource, but each specific resource type has to extend a BaseResource copy metadata method to copy extended metadata elements for each specific resource type. I have spent some time investigating to try to avoid this by directly copying metadate model following django model instance copying doc here https://docs.djangoproject.com/en/1.9/topics/db/queries/#copying-model-instances, but could not succeed due to the complexity of our metadata model since our metadata model involves  inheritance and related objects, which cannot be handled well using a generic model copy method. After consulting with @pkdash and some further investigation on model copying in Django, I went with the current implementation which involves extending a method from BaseResource CoreMetadataElement model to allow copying extended specific metadata elements from each specific resource type. To support resource versioning for reference time series type and web app type, I have to add this copy metadata element method to each of these resource type.

**Comment by dtarb** [comment]
@ironfroggy @alvacouch could you please comment on whether there is a way to simplify @hyi having to implement functionality repeatedly for each resource type.

**Comment by pkdash** [comment]
@hyi You can give the following code a try (untested code):

    def copy_all_elements_to(self, tgt_res):
	   from hydroshare.utils import current_site_url
	   from django.forms.model import model_to_dict
	
	   new_md = tgt_res.metadata
	   metadata_type = ContentType.objects.get_for_model(self)
	   supported_element_names = self.get_supported_element_names()
	   for element_name in supported_element_names:
		 element_model_type = self._get_metadata_element_model_type(element_name)			
		 elements_to_copy = element_model_type.model_class().objects.filter(object_id=self.id, content_type=metadata_type).all()
		 for element in elements_to_copy:
			 new_md.create_element(element_name, model_to_dict(element))

**Comment by hyi** [comment]
@pkdash Thanks for the code snippet. I finally got it working with some modification to the code snippet as well as some changes to how Coverage element is created. This will handle metadata element copying without touching specific resource types. @horsburgh @dtarb @ironfroggy @alvacouch I will deploy my branch to beta for team testing and create a PR for code review soon.

**Comment by pkdash** [comment]
@hyi Glad to know that snippet was helpful.

**Comment by hyi** [comment]
@horsburgh @dtarb I have finally resolved the issue that happened only on beta but not happen in my local development environment. Things appear to work as expected on https://beta.hydroshare.org. Please test the resource versioning and let me know any comments/feedback to address.

**Comment by hyi** [comment]
@horsburgh @dtarb One thing I need to point out when you do testing on beta: since beta is deployed using www resources which has metadata element HydroShare Identifier pointing to www.hydroshare.org site, the URL to the obsoleted resource on the new versioned resource landing page may point to www site. This will not happen on www production site, and is expected for testing on beta.

**Comment by hyi** [comment]
@horsburgh @dtarb Could you let me know whether updated date metadata element needs to be changed for the original resource when a new version has been created for this resource? I think it should since we added "isReplacedBy" relation element to the original resource, so it is indeed modified. If you all agree, the only thing I need to do is to modify updated date dublin core metadata element for the original resource to be in sync with resource updated date set up for the resource page by Mezzanine (which is what is displayed on resource landing page), which is easy to do.

**Comment by horsburgh** [comment]
@hyi - I did some testing of the versioning.  I have the following comments:

- [ ] 1. I would change the order of the button icons at the top.  I would make the order go Edit, Delete, Version, Publish.  It seems weird to have the delete button between version and publish.

- [ ] 2. On the first pop up that appears when you click the Version button, change to text to the following:

"Creating a new version will obsolete this resource. This resource will be still available in HydroShare, but only the newer version will be discoverable via the HydroShare search interface. A link to this resource will be created on the landing page of the new version. Only one newer version is allowed for each resource."

- [ ] 3. Some actions in HydroShare display a success or failure message at the top of the page. Do you want to do this for versioning?  I'm thinking you could display a message at the top in green that says:

"Success! Your new version has been created. A link to the older version has been added below the resource title. To modify this new version or change the sharing status, click the Edit button. To share this new version with individual HydroShare users, click the Manage Access button."

- [ ] 4.  I'm assuming that this will be taken care of with other changes that are being discussed for search, but when I create a new version and then go to the search page, I can discover the old version, but not the new version. This is exactly the opposite of what we want.  I think removal of the old version from the search index and addition of the new version needs to be nearly instantaneous or users will become confused.

- [ ] 5. After creating a new version, the My Resources page shows and provides a link to both versions. In answer to your question above, I would change the updated date metadata element for the obsoleted resource (I'm assuming that is what is displayed in the "Last Modified" column in the table) to reflect that the metadata was changed when the new resource was created. Right now, this is really the only thing that we could use to distinguish between the versions in the table. Another option would be to only display the most recent version in the table on the My Resource page. @dtarb - what do you think about this?

**Comment by hyi** [comment]
@horsburgh Thanks for the comments. I can fix issues No. 1-3 quickly. No. 4 depends on implementation of on-demand search index update @alvacouch and @xuezhaokun would investigate initially as part of issue https://github.com/hydroshare/hydroshare/issues/755 to see whether celery is needed or not, and if not, they can get it done without my involvement, and if celery is needed, I can help. Only after issue https://github.com/hydroshare/hydroshare/issues/755 is resolved and we have on-demand search index update in place, I can add the functionality to remove old version from search and add new version into search instantaneously. I will create an issue for it to keep track of it. For No. 5, I think if we change the updated date metadata element for the original resource, the original resource and the new versioned resource are still indistinguishable since they have the same titles and same last modified date. I like the option to remove the original resource from My Resource page to make it only discoverable from the most recent version via the obsolecence chain, but I'll wait to do it until I hear final words from @dtarb or @horsburgh

**Comment by horsburgh** [comment]
@hyi and @dtarb - in the interest of keeping the My Resources page uncluttered, I vote for only displaying the most recent version of any Resource that has older versions.

**Comment by hyi** [comment]
@horsburgh That is what I prefer as well. I'll go with only displaying the most recent versions on My Resources page unless I hear objections from @dtarb

**Comment by hyi** [comment]
@horsburgh On No. 3 in your comment list, the current implementation only shows that congratulation message when the original resource being versioned is not private, indicating the new version is created as a private resource by default. I have modified the code to show this congratulation message for private resources being versioned as well by combining your suggested message with its existing message as follows:
"Congratulations! Your new version has been created. A link to the older version has been added below the resource title. To modify this new version or change the sharing status, click the Edit button. To share this new version with individual HydroShare users, click the Manage Access button. Note that this new version is created as a private resource by default. If you want the new version to be discoverable, you need to make it public or discoverable or publish it."
Note that I replaced "Success" with "Congratulations" to be consistent with the message we show for other scenarios such as resource just created and resource just published. Let me know if you have any suggestions to the message above.

**Comment by horsburgh** [comment]
@hyi - the message works for me.  Thanks.

**Comment by hyi** [comment]
@horsburgh @dtarb Have addressed all further comments from @horsburgh except for removing obsoleted resources from searches which I have created a separate issue to address after on-demand solr index update for search has been addressed. Also fixed a bug along the way. I have created a PR for code review and hope it can be merged and included in the new deployment.

**Comment by hyi** [comment]
@dtarb @horsburgh BTW, this PR I just created does include removing obsoleted resources from My Resources page, so only the latest version of the resource shows up on My Resources page. This removes confusion and works well in my opinion, but let me know if you have any further suggestions.

**Comment by horsburgh** [comment]
@hyi - Awesome!  I'm excited to see this get merged and deployed!

**Comment by dtarb** [comment]
I too am excited to see this deployed and move on.  One idea that occurred to me is to have an option in myResources to "Show all versions".  DropBox has something like this, but unless this is very simple and natural and quick to implement (and very little seems simple and quick to implement) I would prefer to shelve this idea and move on.  The all hands meeting this week and subsequent discussions with Rick impressed on me the need to address functionality that breaks ahead of these cosmetic things.

**Comment by hyi** [comment]
@dtarb @horsburgh "Show all versions" may be useful on "My Resources" page, but let's shelve this idea and move on at this point. If a need arises doing so in the future, @mauriel133 can take it on to incorporate this into My Resources page.

**Comment by hyi** [comment]
@pkdash I have addressed your code review comments. See if you can +1 now.

**Comment by horsburgh** [comment]
@hyi and @dtarb - Dave had some interesting concerns on the pull request for this issue.  I see Dave’s points about potentially not replacing/obsoleting a resource until the new version has been made public/published, but there’s obviously some complexity there that would have to be dealt with. For example if a new version is made but not published, can I make another new version?  And, if so, how would we manage the version trail?

I think for the first "version" of versioning we should go with what we've got with the warning that when a user creates a new version it removes the old version from discovery right then (whether they make the new version public/published right then or not).

@hyi - this also makes me wonder if you have handled the case where a new version is created and then deleted.  Does the previous version become the version of record again and get added back to data discovery?

**Comment by hyi** [comment]
@horsburgh @dtarb Yes, I did handle the case where a new version is created and then deleted, exactly according to the design spec, i.e., the previous version become the latest version in the version chain again. The data discovery piece is what is missing since it is dependent on the implementation of on-demand solr index update as detailed in issue https://github.com/hydroshare/hydroshare/issues/755. This data discovery missing piece for versioning is recorded in issue https://github.com/hydroshare/hydroshare/issues/1047 which I will tackle after on-demand solr index update is implemented by @alvacouch and/or @xuezhaokun

**Comment by horsburgh** [comment]
Awesome.  Thanks @hyi.

**Comment by dtarb** [comment]
Thanks @hyi.  I am pasting below the gist of the comments I had on the pull request so we do not lose them, but am happy to go with what we have to get some experience with this then adjust based on the experience.

I think we need flexibility regarding the changing of relationships for published resources.  We have said that a published resource is immutable, but I think this should only pertain to the actual data and descriptive metadata.  Metadata relationships such as "superseded by" or "cited by" need to be able to be added and removed from published resources, and should not be regarded as acts of changing the published resource, so should not alter the date used to display the resource in resource listings. This means we have to sort out the datetime fields.  If these changes do change the "last changed" datetime, then resource listings should probably use another datetime field, such as date published.  

Also when a new version of a published resource is created there may be a time lag that needs to be accommodated.  The workflow will most likely be to (1) make a new version, then do some work on it that eventually (perhaps months or years later) results in it being ready to supercede the initially published resource. Only when the new superceding resource is ready, perhaps as indicated by it being "published" should the relationship be added to the old published resource. Until then the published version is the version of record.  The create new version button that immediately results in the resource being superceded does not reflect this time lag.

I would encourage @horsburgh to systematically address these questions with the data publication team to direct future work.

**Comment by hyi** [comment]
Since resource versioning has been included in the latest deployment on www, I am closing this issue. I have already created another issue for removing obsoleted resources from discovery interface. We can create new issues as needed from user experiences once this versioning is being used more and more.
